sheets/batch_get.go

```diff
@@ -0,0 +1,162 @@
+	// spreadsheet ID of the Google sheet
+	spreadsheetID string
+	// gid of the sheet extracted from the sheet URL <url>#gid=<gid>
+	sheetID int64
+	// instance of sheets service, used to interact with Google Sheets APIs
+	sheetSvc *sheets.Service
+	// If rate limit is exceeded, nextRun is used to skip hitting API till the specified time.
+	// Exponential backoff is used to decide nextRun time on rate-limit error, uses retryCount*pollingPeriod seconds as duration
+	nextRun time.Time
+	// polling period defined in the config, it is used to implement exponential backoff
+	pollingPeriod time.Duration
+	// the count of unsuccessful retries made after getting 429(rate-limit exceeded) http error status.
+	retryCount int64
+	// dateTimeRenderOption Determines how dates, times, and durations in the response should be rendered.
+	// This is ignored if responseValueRenderOption is FORMATTED_VALUE.
+	// The default dateTime render option is FORMATTED_STRING for the connector.
+	dateTimeRenderOption string
+	// valueRenderOption Determines how values in the response should be rendered.
+	// The default render option is FORMATTED_VALUE.
+	valueRenderOption string
+}
+
+type BatchReaderArgs struct {
+	OAuthConfig          *oauth2.Config
+	OAuthToken           *oauth2.Token
+	SpreadsheetID        string
+	SheetID              int64
+	DateTimeRenderOption string
+	ValueRenderOption    string
+	PollingPeriod        time.Duration
+}
+
+func NewBatchReader(ctx context.Context, args BatchReaderArgs) (*BatchReader, error) {
+	sheetService, err := sheets.NewService(ctx, option.WithHTTPClient(args.OAuthConfig.Client(ctx, args.OAuthToken)))
+	if err != nil {
+		return nil, fmt.Errorf("error creating sheets service client: %w", err)
+	}
+	return &BatchReader{
+		spreadsheetID:        args.SpreadsheetID,
+		sheetID:              args.SheetID,
+		pollingPeriod:        args.PollingPeriod,
+		sheetSvc:             sheetService,
+		dateTimeRenderOption: args.DateTimeRenderOption,
+		valueRenderOption:    args.ValueRenderOption,
+	}, nil
+}
+
+// GetSheetRecords returns the list of records up to a maximum of 1000 rows(default limit)
+// added after the row offset of last successfully read record
+func (b *BatchReader) GetSheetRecords(ctx context.Context, offset int64) ([]sdk.Record, error) {
+	if b.nextRun.After(time.Now()) {
+		return nil, nil
+	}
+
+	res, err := b.sheetSvc.Spreadsheets.Values.BatchGetByDataFilter(b.spreadsheetID, b.getDataFilter(offset)).Context(ctx).Do()
+	if err != nil {
+		if googleapi.IsNotModified(err) {
+			return nil, nil
+		}
+		if gerr, ok := err.(*googleapi.Error); ok && gerr.Code == http.StatusTooManyRequests {
+			b.retryCount++
+			duration := time.Duration(b.retryCount * int64(b.pollingPeriod)) // exponential back off
+			b.nextRun = time.Now().Add(duration)
+			sdk.Logger(ctx).Error().
+				Int64("retry_count", b.retryCount).
+				Float64("wait_duration", duration.Seconds()).
+				Msg("exponential back off, rate limit exceeded")
```

Comment:
I was actually referring to logging the error itself `gerr`, but that's not a blocker.

Reply:
Added